File: 210-course-schedule-ii/210-course-schedule-ii.ts
function findOrder(numCourses: number, prerequisites: number[][]): number[] {
  return KhanAlgoWithTopologicalSort(numCourses, prerequisites);
}

function DFS_Stack(numCourses: number, prerequisites: number[][]): number[] {
  const adj = Array.from(Array(numCourses), () => []),
        visit = Array(numCourses).fill(false),
        stack = [],
        ans = [];
  
  for(let [s, d] of prerequisites)
    adj[s].push(d);
  
  for(let i = 0; i < numCourses; i++){
    if(hasCycle(i))
      return [];
  }
  
  return ans;
  
  function hasCycle(v: number): boolean {
    if(visit[v]){
      if(stack.includes(v))
        return true;
      
      return false;
    }
    visit[v] = true;
    stack.push(v);
    
    for(let deps of adj[v]){
      if(hasCycle(deps))
        return true;
    }
    
    ans.push(stack.pop());
    
    return false;
  }
};


function KhanAlgoWithTopologicalSort(numCourses: number, prerequisites: number[][]): number[] {
  const adj = Array.from(Array(numCourses), () => []),
        in_degree = Array(numCourses).fill(0),
        order = [],
        q = [];
  let i = numCourses;
    
  for(let [s, d] of prerequisites)
    adj[s].push(d);

  for(let i = 0; i < numCourses; i++)
    for(let deps of adj[i])
      in_degree[deps] += 1; 
  
  // console.log({adj, in_degree})
  
  for(let i = 0; i < numCourses; i++)
    if(in_degree[i] == 0)
      q.unshift(i);

  while(q.length) {
    const at = q.pop();
    order[--i] = at;
    
    for(let deps of adj[at]){
      in_degree[deps] -= 1;
      
      if(in_degree[deps] == 0)
        q.unshift(deps);
    }
  }
    
  // console.log({order, i});
  
  if(i > 0)
    return [];

  return order;
  
}

File: 210-course-schedule-ii/210-course-schedule-ii.java
class Solution {
    private boolean[] visited;
    private List<Integer> ans;
    private Map<Integer, List<Integer>> graph;
    private int[] inDegree;
    
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        this.visited = new boolean[numCourses];
        this.ans = new ArrayList<>();
        this.inDegree = new int[numCourses];
        this.buildGraph(numCourses, prerequisites);
        
        Queue<Integer> q = new LinkedList<>();
        
        for(int i = 0; i < numCourses; i++)
            for(int deps: this.graph.get(i))
                this.inDegree[deps] += 1;
            
        for(int i = 0; i < numCourses; i++)
            if(this.inDegree[i] == 0)
                q.offer(i);
        
        while(!q.isEmpty()) {
            int at = q.poll();
            this.ans.add(at);
            
            for(int deps: this.graph.get(at)) {
                this.inDegree[deps] -= 1;
                
                if(this.inDegree[deps] == 0)
                    q.offer(deps);
            }
        }
        
        if(this.ans.size() != numCourses)
            return new int[]{};
        
        int[] res = new int[this.ans.size()];
        
        for(int i = this.ans.size() - 1; i >= 0; i--)
            res[i] = this.ans.get(this.ans.size() - 1 - i);
        
        return res;
    }
    
    private void buildGraph(int N, int[][] prerequisites) {
        this.graph = new HashMap<>();
        
        while(N-- > 0)
            this.graph.put(N, new ArrayList<>());
        
        for(int[] conn: prerequisites)
            this.graph.get(conn[0]).add(conn[1]);
    }
}